File: frontend/src/app/core/services/message.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { ApiService } from './api.service';
import { ConversationResponse, MessageResponse } from '../models/message.model';
import { PageResponse } from '../models/api.model';

@Injectable({ providedIn: 'root' })
export class MessageService {
  private api = inject(ApiService);

  getOrCreateConversation(contractId: number): Observable<ConversationResponse> {
    return this.api.get<ConversationResponse>(`/contracts/${contractId}/conversation`);
  }

  getMessages(conversationId: number, page = 0, size = 50): Observable<PageResponse<MessageResponse>> {
    return this.api.get<PageResponse<MessageResponse>>(
      `/conversations/${conversationId}/messages`, { page, size }
    );
  }

  sendMessage(conversationId: number, content: string, attachmentUrl?: string): Observable<MessageResponse> {
    return this.api.post<MessageResponse>(`/conversations/${conversationId}/messages`, { content, attachmentUrl });
  }
}

File: backend/src/main/java/com/freelancer/service/MessageService.java
package com.freelancer.service;

import com.freelancer.dto.request.SendMessageRequest;
import com.freelancer.dto.response.ConversationResponse;
import com.freelancer.dto.response.MessageResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface MessageService {
    ConversationResponse getOrCreateConversation(Long contractId, Long userId);
    Page<MessageResponse> getMessages(Long conversationId, Long userId, Pageable pageable);
    MessageResponse sendMessage(Long conversationId, Long senderId, SendMessageRequest req);
}
